File: scripts/update_version.py
#!/usr/bin/env python3
"""
Flutter Leaf Kit - Version Update Script

Usage:
    update-version                      # 대화형으로 버전 입력
    update-version 2.5.0                # 버전 직접 지정
    update-version 2.5.0 --auto-commit  # 자동 커밋
"""

import os
import argparse
import ruamel.yaml
from pathlib import Path
from git import Repo
from git.exc import GitCommandError


def get_project_root() -> Path:
    """프로젝트 루트 디렉토리 반환"""
    # 스크립트 위치 기준으로 상위 디렉토리
    return Path(__file__).parent.parent


def main():
    parser = argparse.ArgumentParser(description="Flutter Leaf Kit - Version Update Script")
    parser.add_argument("version", nargs="?", help="버전 (예: 2.5.0)")
    parser.add_argument("--auto-commit", action="store_true", default=False, help="자동 커밋 (기본값: false)")
    args = parser.parse_args()

    yaml = ruamel.yaml.YAML()
    yaml.preserve_quotes = True

    project_root = get_project_root()
    os.chdir(project_root)

    # 버전 입력 받기
    print("📦 Flutter Leaf Kit - Version Update")
    print()
    if args.version:
        update_version = args.version
        print(f"🔢 Version: {update_version}")
    else:
        update_version = input("🔢 Please enter the updated version. (ex., 1.0.0): ")

    results = list(Path("./packages").rglob("pubspec.yaml"))

    # Get Sub Package Names
    package_names = list()
    for result in results:
        file_path = str(result)
        package_name = "flutter_{}".format(
            file_path.replace("packages/", "").replace("/pubspec.yaml", "")
        )
        if "flutter_leaf_" in package_name:
            package_names.append(package_name)

    # Update Yaml
    for result in results:
        file_path = str(result)

        with open(file_path, "r") as stream:
            data = yaml.load(stream)

            data["version"] = update_version

            for sub_package_name in package_names:
                try:
                    data["dependencies"][sub_package_name]["git"][
                        "ref"
                    ] = f"v{update_version}"
                except:
                    continue

        with open(file_path, "wb") as stream:
            yaml.dump(data, stream)

    print("\n📝 Yaml files have been modified.")

    # 현재 브랜치가 develop인지 확인
    repo = Repo(project_root)
    current_branch = repo.active_branch.name
    if current_branch != "develop":
        print(f"\n❌ Current branch is '{current_branch}'. This script must be run on the 'develop' branch.")
        return

    # Git commit 여부 확인
    print()
    print("=" * 50)
    print("🔄 Git Commit")
    print(f"   Message: chore: Update version v{update_version}")
    print("=" * 50)

    if args.auto_commit:
        commit_confirm = "y"
        print("Auto commit enabled.")
    else:
        commit_confirm = input("Do you want to commit? (y/n): ").strip().lower()

    if commit_confirm == "y":
        commit_message = f"chore: Update version v{update_version}"
        tag_name = f"v{update_version}"

        try:
            # Stage all changes
            repo.git.add("-A")

            # Commit
            repo.index.commit(commit_message)
            print(f"\n✅ Committed: {commit_message}")

            # Git push
            print("\n📤 Pushing to remote...")
            origin = repo.remote("origin")
            origin.push()
            print("✅ Push successful.")

            # Create tag
            print(f"\n🏷️  Creating tag: {tag_name}")
            repo.create_tag(tag_name)
            print(f"✅ Tag created: {tag_name}")

            # Push tag to remote
            print(f"\n📤 Pushing tag to remote...")
            origin.push(tag_name)
            print(f"✅ Tag pushed: {tag_name}")

            print("\n" + "=" * 50)
            print("🎉 Version update completed!")
            print("=" * 50)

            # develop -> main 머지 및 푸시
            print("\n🔀 Merging develop into main...")
            repo.git.checkout("main")
            repo.git.merge("develop")
            print("✅ Merge successful.")

            print("\n📤 Pushing main to remote...")
            origin.push()
            print("✅ Main push successful.")

            print("\n🔄 Switching back to develop branch...")
            repo.git.checkout("develop")
            print("✅ Back on develop branch.")

        except GitCommandError as e:
            print(f"\n❌ Git error: {e}")
    else:
        print("\n⏭️  Commit skipped.")


if __name__ == "__main__":
    main()

File: scripts/update-version.js
#!/usr/bin/env node
/**
 * Flutter Leaf Kit - Version Update Script
 *
 * Usage:
 *   yarn update-version                       # 대화형으로 버전 입력
 *   yarn update-version 2.5.0                 # 버전 직접 지정
 *   yarn update-version 2.5.0 --auto-commit   # 자동 커밋
 */

import { readFile, writeFile, readdir, stat } from 'node:fs/promises';
import { join, dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { parseArgs, promisify } from 'node:util';
import { createInterface } from 'node:readline/promises';
import { stdin as input, stdout as output } from 'node:process';
import { execFile as execFileCb } from 'node:child_process';
import process from 'node:process';
import { parseDocument } from 'yaml';

const runFile = promisify(execFileCb);
const projectRoot = resolve(dirname(fileURLToPath(import.meta.url)), '..');

async function git(...args) {
  const { stdout } = await runFile('git', args, { cwd: projectRoot });
  return stdout.trim();
}

async function fileExists(path) {
  try {
    await stat(path);
    return true;
  } catch {
    return false;
  }
}

async function findPackagePubspecs() {
  const packagesDir = join(projectRoot, 'packages');
  const entries = await readdir(packagesDir, { withFileTypes: true });
  const results = [];
  for (const entry of entries) {
    if (!entry.isDirectory()) continue;
    const pubspecPath = join(packagesDir, entry.name, 'pubspec.yaml');
    if (await fileExists(pubspecPath)) results.push(pubspecPath);
  }
  return results.sort();
}

function setScalar(doc, path, value) {
  const node = doc.getIn(path, true);
  if (node && typeof node === 'object' && 'value' in node) {
    node.value = value;
    return true;
  }
  return false;
}

async function ask(question) {
  const rl = createInterface({ input, output });
  try {
    return await rl.question(question);
  } finally {
    rl.close();
  }
}

async function main() {
  const { values, positionals } = parseArgs({
    options: {
      'auto-commit': { type: 'boolean', default: false },
    },
    allowPositionals: true,
  });

  const pubspecFiles = await findPackagePubspecs();

  console.log('📦 Flutter Leaf Kit - Version Update');
  console.log('');

  let updateVersion = positionals[0];
  if (updateVersion) {
    console.log(`🔢 Version: ${updateVersion}`);
  } else {
    updateVersion = (await ask('🔢 Please enter the updated version. (ex., 1.0.0): ')).trim();
  }

  if (!updateVersion) {
    console.log('\n❌ Version is required.');
    process.exit(1);
  }

  const subPackageNames = pubspecFiles
    .map((p) => 'flutter_' + p.replace(projectRoot + '/', '').replace('packages/', '').replace('/pubspec.yaml', ''))
    .filter((name) => name.includes('flutter_leaf_'));

  for (const filePath of pubspecFiles) {
    const content = await readFile(filePath, 'utf-8');
    const doc = parseDocument(content);

    setScalar(doc, ['version'], updateVersion);

    for (const subPkg of subPackageNames) {
      setScalar(doc, ['dependencies', subPkg, 'git', 'ref'], `v${updateVersion}`);
    }

    await writeFile(filePath, doc.toString());
  }

  console.log('\n📝 Yaml files have been modified.');

  const currentBranch = await git('rev-parse', '--abbrev-ref', 'HEAD');
  if (currentBranch !== 'develop') {
    console.log(`\n❌ Current branch is '${currentBranch}'. This script must be run on the 'develop' branch.`);
    return;
  }

  console.log('');
  console.log('='.repeat(50));
  console.log('🔄 Git Commit');
  console.log(`   Message: chore: Update version v${updateVersion}`);
  console.log('='.repeat(50));

  let commitConfirm;
  if (values['auto-commit']) {
    commitConfirm = 'y';
    console.log('Auto commit enabled.');
  } else {
    commitConfirm = (await ask('Do you want to commit? (y/n): ')).trim().toLowerCase();
  }

  if (commitConfirm !== 'y') {
    console.log('\n⏭️  Commit skipped.');
    return;
  }

  const commitMessage = `chore: Update version v${updateVersion}`;
  const tagName = `v${updateVersion}`;

  try {
    await git('add', '-A');
    await git('commit', '-m', commitMessage);
    console.log(`\n✅ Committed: ${commitMessage}`);

    console.log('\n📤 Pushing to remote...');
    await git('push', 'origin', 'develop');
    console.log('✅ Push successful.');

    console.log(`\n🏷️  Creating tag: ${tagName}`);
    await git('tag', tagName);
    console.log(`✅ Tag created: ${tagName}`);

    console.log('\n📤 Pushing tag to remote...');
    await git('push', 'origin', tagName);
    console.log(`✅ Tag pushed: ${tagName}`);

    console.log('');
    console.log('='.repeat(50));
    console.log('🎉 Version update completed!');
    console.log('='.repeat(50));

    console.log('\n🔀 Merging develop into main...');
    await git('checkout', 'main');
    await git('merge', 'develop');
    console.log('✅ Merge successful.');

    console.log('\n📤 Pushing main to remote...');
    await git('push', 'origin', 'main');
    console.log('✅ Main push successful.');

    console.log('\n🔄 Switching back to develop branch...');
    await git('checkout', 'develop');
    console.log('✅ Back on develop branch.');
  } catch (e) {
    console.log(`\n❌ Git error: ${e.stderr?.toString() || e.message}`);
    process.exit(1);
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
